Feature/runtime execution engine

AgustinGil21 opened the pull request:
This draft adds the following implementations to the Linux runtime:

Execute, ExecuteWithTimeout, ExecuteWithEnvironment, StartProcess, StopProcess, KillProcess, GetProcessStatus, ListProcesses, CaptureOutput, CaptureError, StreamOutput, StreamError, CleanupProcess, CleanupAllProcesses, and Shutdown.

Additionally, corresponding tests were added to cover each implementation.
These tests are not tied to the operating system, ensuring cross-platform compatibility.

AgustinGil21 committed:
OS runtime implementations

AgustinGil21 committed:
Minor changes

AgustinGil21 committed:
Minor changes

AgustinGil21 committed:
Linux runtime implementations

AgustinGil21 committed:
Code format

AgustinGil21 committed:
Some new linux runtime implementations

AgustinGil21 committed:
Linux runtime ListProcesses implementation improvements

AgustinGil21 committed:
Linux runtime new implementation and some other changes

AgustinGil21 committed:
Tests rewritten and some other minor changes

AgustinGil21 committed:
Tests fixes

AgustinGil21 committed:
Format code

char2cs commented on internal/infrastructure/runtime/runtime.go:
```diff
@@ -1,28 +1,91 @@
+	REEImplementation REEInterface
 }
 
+var CurrentOS system.OS
```
No recomiendo para nada esto, podemos guardarlo en `Runtime` que es arquitectonicamente bien, y no creamos variables globales. Tratemos de evitarlas -> Dificultan mucho los tests

char2cs commented on internal/infrastructure/runtime/darwin-runtime-implementation.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+func (r *DarwinRuntime) Execute(ctx context.Context, path string, args []string) (string, error) {
+	// execute command context
```
Estos `comments` me parecen redundantes, se sobreentiende mirando el codigo

char2cs commented on internal/infrastructure/runtime/runtime.go:
```diff
@@ -1,28 +1,91 @@
+		return &LinuxRuntime{Runtime: &Runtime{}, processes: make(map[string]*ProcessInfo)}
+	} else if CurrentOS.IsDarwin() {
+		return &DarwinRuntime{Runtime: &Runtime{}}
+	}
```
Podemos implementar un Factory Pattern (como esta en el `netbridge`) para que sea mas claro y facil agregar nuevas implementaciones.

AgustinGil21 replied:
Falta por implementar

char2cs commented on internal/infrastructure/runtime/linux-runtime-implementation.go:
```diff
@@ -0,0 +1,478 @@
+	ctx context.Context,
+	command []string,
+	env map[string]string,
+) (string, error) {
```
Evaluemos devolver un string aca, capaz es mas util devolver el objeto del proceso. 

(Propongo un refactor) -> Estaria bueno poder devolver un objeto proceso, y llamando a metodos del proceso poder controlarlo, por fuera del runtime, serian sus propias entidades que implementan internamente una implementacion de OS (Darwin, Linux o Windows). 

Me imagino que seria algo asi como lo usariamos de afuera
```go
process, err := runtime.Get(ctx, arrow.executable.path, []string)

process.Start()
process.Stop()
process.Kill()
etc...
```

AgustinGil21 replied:
Falta por implementar

char2cs commented on internal/infrastructure/runtime/linux-runtime-implementation.go:
```diff
@@ -0,0 +1,478 @@
+		// close channels
+		close(info.OutChan)
+		close(info.ErrorChan)
+	}()
```
Se puede dar el caso de que haya un leak de goroutine aca.

Si nunca finaliza el canal de info.OutChan -> Podria pasar que esta goroutine nunca finalice, aun cuando haya terminado el context. Mi recomendacion agregar un select -> Usando el ctx.

```go
select {
case err := <-done:
      // handle exit
case <-ctx.Done():
      cmd.Process.Kill()
      // handle cancellation
}
```

char2cs commented on internal/infrastructure/runtime/linux-runtime-implementation.go:
```diff
@@ -0,0 +1,478 @@
+				return nil // finished
+			}
+		}
+	}
```
Esto es ineficiente, estaria preguntando cada 100ms si el proceso finalizo o no. Podriamos hacerlo lazy, lanzar una goroutine y rezarle a dios de que finalice. 

O mismo dejarlo a razon de la API, si un proceso no finaliza por `X` razon, es reponsabilidad del usuario si quiere forzar de que se cierre (Usariamos Kill en ese caso).

char2cs commented on internal/infrastructure/runtime/linux-runtime-implementation.go:
```diff
@@ -0,0 +1,478 @@
+	process.Status = "killing"
+	r.processLock.Unlock()
+
+	// wait for monitor to set finished
```
Esta goroutine saldria antes de que podamos terminar de matar el proceso, tiene un timeout muy cortito

char2cs commented on internal/infrastructure/runtime/linux-runtime-implementation.go:
```diff
@@ -0,0 +1,478 @@
+	}
+
+	// save process
+	for uuid := range r.processes {
```
Guarda con llamar variables de la misma forma que `gomodules`, podriamos llamarlo id, uid, pero nunca el mismo nombre que el modulo -> Seria shadowing.

char2cs commented on internal/infrastructure/runtime/linux-runtime-implementation.go:
```diff
@@ -0,0 +1,478 @@
+
+		if err != nil {
+			return err
+		}
```
Podria fallar a mitad de recorrer el array, provocando que algunos process no se borren correctamente

char2cs reviewed (changes requested):
Sacando estas cosas Agus esta impecable. Miremos, si queres, el finde el tema del objeto, me gustaria y creo que seria mas comodo que cada objeto `process` maneje su propio `environment`. Cada uno seria responsable de arrancar, parar y stoppearse a si mismos. 

Internamente implementarian el mismo objeto del REE y estos objetos, pero se expondria de otra forma, no se si me explico.

AgustinGil21 committed:
Linux runtime implementation improvements

AgustinGil21 committed:
Darwin and Windows runtime implementations

AgustinGil21 committed:
Linux, Darwin and Windows runtime tests

AgustinGil21 committed:
Linux runtime implementation

AgustinGil21 committed:
Darwin implementation fixes and tests fully rewritten

AgustinGil21 committed:
Windows runtime implementation fixes and windows tests fully rewritten

AgustinGil21 committed:
Linux runtime tests minor changes

AgustinGil21 committed:
Minor changes

AgustinGil21 committed:
Linux runtime tests minor fixe

AgustinGil21 committed:
Linux runtime new tests

AgustinGil21 committed:
Linux runtime ExecuteWithEnvironment changes

AgustinGil21 committed:
Linux runtime ExecuteWithTimeout improvements

AgustinGil21 committed:
Runtime new test NewRuntime

AgustinGil21 committed:
Runtime minor changes

AgustinGil21 committed:
Runtime new tests

AgustinGil21 committed:
Minor changes

char2cs committed:
refactor(runtime): update runtime initialization and error handling

- Changed Runtime field in Infrastructure struct to a pointer.
- Enhanced error handling during runtime initialization to allow application functionality without runtime support.
- Removed unused Darwin and Linux runtime implementations and their associated tests to streamline the codebase.

char2cs committed:
refactor(errors): remove unnecessary blank line in error definitions

char2cs committed:
refactor(tests): remove unnecessary blank lines in test files

char2cs committed:
refactor(runtime): rename IsTerminal to IsFinished and enhance output logging

- Updated the Status method from IsTerminal to IsFinished for clarity.
- Added logging for dropped lines in output channels to improve monitoring.
- Introduced a new method in Builder to set configuration, enhancing flexibility.
- Increased timeout duration for process termination across all platforms to ensure graceful shutdown.

char2cs committed:
refactor(tests): enhance netbridge and runtime tests for better coverage

- Added new test cases for port availability and forwarding in the netbridge implementation.
- Improved test structure for the runtime manager, including concurrent access and error handling scenarios.
- Updated existing tests to ensure they accurately reflect the current implementation and improve overall test coverage.
- Renamed test functions for clarity and consistency.

AgustinGil21 committed:
Merge pull request #66 from rabbytesoftware/feature/ree-refactor

refactor/runtime

AgustinGil21 committed:
Linux tests fixes

Lucacux committed:
Merge branch 'develop' into feature/runtime-execution-engine

Lucacux reviewed (approved):
Great work on this PR. The introduction of the Runtime Execution Engine is well structured and cleanly organized across OS-specific implementations. The builder pattern for process creation makes the API very readable and extensible, and the use of context.Context throughout is solid.

The improvements to the test suite—especially around netbridge behavior, interface compliance, and consistent return values—add a lot of reliability to the codebase.

Overall, the changes are well designed, easy to follow, and strengthen both the runtime layer and test coverage.
LGTM — Approved.
